pallets/contracts2/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+    macro_rules! on {
+        ($p:pat, $e:pat) => {
+            FuncId {
+                scheme: 0,
+                pallet: $p,
+                extrinsic: $e,
+                version: 0,
+            }
+        };
+    }
+
+    // TODO(Centril): Should we charge weight depending on branch?
+    // That would require us to benchmark the decoding of each and every call.
+
+    Ok(match func_id {
+        on!(0, 0) => CommonCall::Asset(pallet_asset::Call::register_ticker { ticker: decode!() }),
+        on!(0, 1) => {
+            CommonCall::Asset(pallet_asset::Call::accept_ticker_transfer { auth_id: decode!() })
+        }
+        on!(0, 2) => CommonCall::Asset(pallet_asset::Call::accept_asset_ownership_transfer {
+            auth_id: decode!(),
+        }),
+        on!(0, 3) => CommonCall::Asset(pallet_asset::Call::create_asset {
+            name: decode!(),
+            ticker: decode!(),
+            divisible: decode!(),
+            asset_type: decode!(),
+            identifiers: decode!(),
+            funding_round: decode!(),
+            disable_iu: decode!(),
+        }),
+        _ => return Err(Error::<T>::RuntimeCallNotFound.into()),
+    })
+}
+
+/// A chain extension allowing calls to polymesh pallets
+/// and using the contract's DID instead of the caller's DID.
+impl<T> ce::ChainExtension<T> for Module<T>
+where
+    <T as pallet_contracts::Config>::Call: From<CommonCall<T>> + GetDispatchInfo,
+    T: Config + pallet_asset::Config,
+{
+    fn enabled() -> bool {
+        true
+    }
+
+    fn call<E: ce::Ext<T = T>>(
+        func_id: u32,
+        env: ce::Environment<E, ce::InitState>,
+    ) -> ce::Result<ce::RetVal> {
+        let mut env = env.buf_in_buf_out();
+
+        // TODO(Centril): charge weight + benchmark code in here?
+        // It becomes a bit cyclical,
+        // as charging weight itself becomes part of the computation to charge weight for.
+
+        // Swap current DID to caller's DID.
+        let old_did = set_current_did_to_key::<T>(env.ext().caller());
+
+        // Decide what to call in the runtime.
+        let func_id = split_func_id(func_id);
+        // TODO(Centril): charge weight + benchmark depending on `in_len`.
+        // Also, should we impose a max limit on `in_len`?
+        let input = &mut &*env.read(env.in_len())?;
+        let call: <T as pallet_contracts::Config>::Call =
+            construct_call::<T>(func_id, input)?.into();
+        ensure!(input.is_empty(), Error::<T>::DataLeftAfterDecoding);
+
+        // Charge weight for the call.
+        let di = call.get_dispatch_info();
+        let charged_amount = env.charge_weight(di.weight)?;
+
+        // Execute call requested by contract.
+        let result = env.ext().call_runtime(call);
+
+        // Refund unspent weight.
+        let post_di = result.unwrap_or_else(|e| e.post_info);
+        let actual_weight = post_di.calc_actual_weight(&di);
+        env.adjust_weight(charged_amount, actual_weight);
```

Comment:
Only need to call `env.adult_weight()` if `post_di.actual_weight.is_some()`.  `post_di.calc_actual_weight(&di)` will just return `di.weight` if the extrinsic doesn't return the actual weight (which is what all of our extrinsics currently do).

Reply:
There are some extrinsics that use `actual_weight`. E.g `relay_tx`.

Reply:
There are some extrinsics that use `actual_weight`. E.g `relay_tx`.

Reply:
`relay_tx` just returns the actual weights from the relay call (plus some fixed cost), which is good for the future when we start returning actual weights from our other extrinsics.  We should do the same for the other `batch*` extrinsics.

Reply:
N.B. `adjust_weight` does nothing if the charged amount and actual weights are the same, but checking `.is_some()` likely does less work.